File: src/api/swagger.ts
import apiClient from './client';

// ========== ALERT MANAGEMENT ==========

export interface Alert {
  id: string;
  alert_name: string;
  severity: string;
  status: string;
  processing_status: string;
  summary: string;
  description: string;
  starts_at: string;
  ends_at?: string;
  fingerprint: string;
  labels: Record<string, string>;
  contact_points_processed_at?: string;
  created_at: string;
  updated_at: string;
}

export const alertsService = {
  listAlerts: (limit = 50, offset = 0, filters?: Record<string, string>) =>
    apiClient.get<{ status: string; total: number; data: Alert[] }>('/alerts', {
      params: { limit, offset, ...filters },
    }),

  getAlert: (id: string) =>
    apiClient.get<{ status: string; data: Alert }>(`/alerts/${id}`),

  getStats: () =>
    apiClient.get<any>('/alerts/stats'),
};

// ========== CONTACT POINTS ==========

export interface LabelMapping {
  key: string;
  operator?: string;
  value?: string;
}

export interface ContactPoint {
  id: string;
  name: string;
  type: string;
  config: Record<string, any>;
  match_logic: string;
  label_mappings: LabelMapping[];
  ignore_matchers: LabelMapping[];
  group_name?: string;
  priority: number;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

export interface ContactPointCreate {
  name: string;
  type: string;
  config: Record<string, any>;
  match_logic?: string;
  label_mappings?: LabelMapping[];
  ignore_matchers?: LabelMapping[];
  group_name?: string;
  priority?: number;
  enabled?: boolean;
}

export const contactPointsService = {
  create: (data: ContactPointCreate) =>
    apiClient.post<ContactPoint>('/api/contact-points', data),

  list: (enabledOnly = false) =>
    apiClient.get<ContactPoint[]>('/api/contact-points', {
      params: { enabled_only: enabledOnly },
    }),

  get: (id: string) =>
    apiClient.get<ContactPoint>(`/api/contact-points/${id}`),

  getByName: (name: string) =>
    apiClient.get<ContactPoint>(`/api/contact-points/name/${name}`),

  update: (id: string, data: Partial<ContactPoint>) =>
    apiClient.put<ContactPoint>(`/api/contact-points/${id}`, data),

  delete: (id: string) =>
    apiClient.delete(`/api/contact-points/${id}`),

  enable: (id: string) =>
    apiClient.post<ContactPoint>(`/api/contact-points/${id}/enable`),

  disable: (id: string) =>
    apiClient.post<ContactPoint>(`/api/contact-points/${id}/disable`),
};

// ========== VM MAPPINGS ==========

export interface VMMapping {
  id: string;
  vm_id?: string;
  vm_name?: string;
  vm_private_ip?: string;
  owner?: string;
  created_at: string;
  updated_at: string;
}

export const vmMappingsService = {
  create: (data: VMMapping) =>
    apiClient.post<VMMapping>('/api/vm-mappings', data),

  list: () =>
    apiClient.get<VMMapping[]>('/api/vm-mappings'),

  get: (id: string) =>
    apiClient.get<VMMapping>(`/api/vm-mappings/${id}`),

  update: (id: string, data: Partial<VMMapping>) =>
    apiClient.put<VMMapping>(`/api/vm-mappings/${id}`, data),

  delete: (id: string) =>
    apiClient.delete(`/api/vm-mappings/${id}`),
};

// ========== ESCALATION MAPPINGS ==========

export interface EscalationMapping {
  id: string;
  business_line: string;
  l1?: string;
  l2?: string;
  created_at: string;
  updated_at: string;
}

export const escalationMappingsService = {
  create: (data: EscalationMapping) =>
    apiClient.post<EscalationMapping>('/api/escalation-mappings', data),

  list: () =>
    apiClient.get<EscalationMapping[]>('/api/escalation-mappings'),

  get: (id: string) =>
    apiClient.get<EscalationMapping>(`/api/escalation-mappings/${id}`),

  update: (id: string, data: Partial<EscalationMapping>) =>
    apiClient.put<EscalationMapping>(`/api/escalation-mappings/${id}`, data),

  delete: (id: string) =>
    apiClient.delete(`/api/escalation-mappings/${id}`),
};

// ========== EXTRACTION RULES ==========

export interface ExtractionRule {
  id: string;
  name: string;
  label_key: string;
  regex_pattern: string;
  output_key: string;
  description?: string;
  enabled: boolean;
  priority: number;
  created_at: string;
  updated_at: string;
}

export const extractionRulesService = {
  create: (data: ExtractionRule) =>
    apiClient.post<ExtractionRule>('/api/extraction-rules', data),

  list: (enabledOnly = false) =>
    apiClient.get<ExtractionRule[]>('/api/extraction-rules', {
      params: { enabled_only: enabledOnly },
    }),

  get: (id: string) =>
    apiClient.get<ExtractionRule>(`/api/extraction-rules/${id}`),

  update: (id: string, data: Partial<ExtractionRule>) =>
    apiClient.put<ExtractionRule>(`/api/extraction-rules/${id}`, data),

  delete: (id: string) =>
    apiClient.delete(`/api/extraction-rules/${id}`),
};

File: pkg/models/swagger.go
package models

import "time"

// ========== ALERT MANAGEMENT ==========

// AlertPayload is the incoming alert from Grafana webhook
type AlertPayload struct {
	Status string         `json:"status"`
	Alerts []AlertDetail  `json:"alerts"`
}

// AlertDetail represents a single alert from Grafana
type AlertDetail struct {
	Status      string            `json:"status"`
	Labels      map[string]string `json:"labels"`
	Annotations map[string]string `json:"annotations"`
	StartsAt    time.Time         `json:"starts_at"`
	EndsAt      time.Time         `json:"ends_at"`
	Fingerprint string            `json:"fingerprint"`
}

// AlertLog represents stored alert in database
type AlertLog struct {
	ID                      string                 `json:"id"`
	AlertName               string                 `json:"alert_name"`
	Severity                string                 `json:"severity"`
	Status                  string                 `json:"status"` // firing, resolved
	ProcessingStatus        string                 `json:"processing_status"` // pending, sent, failed
	Summary                 string                 `json:"summary"`
	Description             string                 `json:"description"`
	StartsAt                time.Time              `json:"starts_at"`
	EndsAt                  *time.Time             `json:"ends_at"`
	Fingerprint             string                 `json:"fingerprint"`
	Labels                  map[string]string      `json:"labels"`
	ContactPointsProcessedAt *time.Time             `json:"contact_points_processed_at"`
	ErrorMessage            string                 `json:"error_message"`
	Retries                 int                    `json:"retries"`
	AlertJSON               string                 `json:"alert_json"`
	CreatedAt               time.Time              `json:"created_at"`
	UpdatedAt               time.Time              `json:"updated_at"`
}

// AlertStatsResponse is stats response
type AlertStatsResponse struct {
	Status            string `json:"status"`
	AlertLogs         map[string]int `json:"alert_logs"`
	ProcessingQueue   map[string]int `json:"processing_queue"`
	Timestamp         time.Time `json:"timestamp"`
}

// ========== CONTACT POINTS ==========

// LabelMapping is for routing alerts
type LabelMapping struct {
	Key       string `json:"key"`
	Operator  string `json:"operator"` // =, !=, =~, !~
	Value     string `json:"value"`
}

// ContactPointCreate is request to create contact point
type ContactPointCreate struct {
	Name           string          `json:"name"`
	Type           string          `json:"type"` // telegram, slack, webhook, email
	Config         map[string]interface{} `json:"config"`
	MatchLogic     string          `json:"match_logic"` // ALL, ANY
	LabelMappings  []LabelMapping  `json:"label_mappings"`
	IgnoreMatchers []LabelMapping  `json:"ignore_matchers"`
	GroupName      *string         `json:"group_name"`
	Priority       int             `json:"priority"`
	Enabled        bool            `json:"enabled"`
}

// ContactPoint represents a contact point in database
type ContactPoint struct {
	ID             string         `json:"id"`
	Name           string         `json:"name"`
	Type           string         `json:"type"`
	Config         map[string]interface{} `json:"config"`
	MatchLogic     string         `json:"match_logic"`
	LabelMappings  []LabelMapping `json:"label_mappings"`
	IgnoreMatchers []LabelMapping `json:"ignore_matchers"`
	GroupName      *string        `json:"group_name"`
	Priority       int            `json:"priority"`
	Enabled        bool           `json:"enabled"`
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
}

// ========== DATA ENRICHMENT ==========

// VMMapping for enriching alerts with VM info
type VMMapping struct {
	ID           string    `json:"id"`
	VMID         *string   `json:"vm_id"`
	VMName       *string   `json:"vm_name"`
	VMPrivateIP  *string   `json:"vm_private_ip"`
	Owner        *string   `json:"owner"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}

// EscalationMapping for L1/L2 support
type EscalationMapping struct {
	ID           string    `json:"id"`
	BusinessLine string    `json:"business_line"`
	L1           *string   `json:"l1"`
	L2           *string   `json:"l2"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}

// ExtractionRule for regex-based label extraction
type ExtractionRule struct {
	ID           string    `json:"id"`
	Name         string    `json:"name"`
	LabelKey     string    `json:"label_key"`
	RegexPattern string    `json:"regex_pattern"`
	OutputKey    string    `json:"output_key"`
	Description  *string   `json:"description"`
	Enabled      bool      `json:"enabled"`
	Priority     int       `json:"priority"`
	CreatedAt    time.Time `json:"created_at"`
	UpdatedAt    time.Time `json:"updated_at"`
}

// ========== RESPONSE MODELS ==========

// WebhookResponse is response to webhook
type WebhookResponse struct {
	Status       string    `json:"status"`
	AlertsCount  int       `json:"alerts_count"`
	Timestamp    time.Time `json:"timestamp"`
}

// HealthCheckResponse is for health endpoint
type HealthCheckResponse struct {
	Status  string `json:"status"`
	Message string `json:"message"`
}

// ErrorResponse is generic error response
type ErrorResponse struct {
	Status  string `json:"status"`
	Error   string `json:"error"`
	Message string `json:"message"`
}
